### ISSUE TITLE
"Unnecessary dict call" false positives

### ISSUE THREAD

**Issue by faulkner** [adamchainz/flake8-comprehensions#64]
The dict function has 3 forms: `dict(**kwarg)`, `dict(mapping, **kwarg)`, and `dict(iterable, **kwarg)`.  I'm not sure if it makes sense to apply this rule to every form.

Here's one example use case:
```py
def foo(data):
    defaults = {'a': 1, 'b': 2}
    return dict(defaults, **data)
```

This example could be rewritten to use literals, but it would be less succinct and (I assume, but haven't confirmed) less performant.

**Comment by adamchainz** [comment]
Are you really getting a false positive? Because C408 checks if the call has no arguments, i.e. it should only report for `dict()`

**Comment by amateja** [comment]
I also confirm what @faulkner has reported.

**Comment by adamchainz** [comment]
I tried to replicate it with a test, both using your example @faulkner and also with a more minimal version (below) but both passed fine:

```patch
diff --git a/tests/test_it.py b/tests/test_it.py
index de600c8..cff7b67 100644
--- a/tests/test_it.py
+++ b/tests/test_it.py
@@ -350,6 +350,15 @@ def test_C408_pass_4():
     assert errors == []
 
 
+def test_C408_pass_5():
+    errors = run_flake8('''\
+        foo = {}
+        bar = {'a': 1}
+        baz = dict(foo, **bar)
+    ''')
+    assert errors == []
+
+
 def test_C408_fail_1():
     errors = run_flake8('tuple()')
     assert errors == [

```

Can either of you provide a test case?

**Comment by amateja** [comment]
I tried a "manual" way:
1. Create empty python3.6.1 virtualenv.
`$ mkvirtualenv -p python3 flake`
2. Install flake8-comprehensions.
`pip install flake8-comprehensions`
3. Verify result.
```
$ pip list --format columns
Package               Version
--------------------- -------
appdirs               1.4.3  
flake8                3.3.0  
flake8-comprehensions 1.4.0  
mccabe                0.6.1  
packaging             16.8   
pip                   9.0.1  
pycodestyle           2.3.1  
pyflakes              1.5.0  
pyparsing             2.2.0  
setuptools            35.0.2 
six                   1.10.0 
wheel                 0.29.0 
```
4. Create an example.py file.
```
echo "a = {'foo': 'bar'}
> dict(baz='bar', **a)" > example.py
```
5. Run flake8.
```
$ flake8 example.py 
example.py:2:1: C408 Unnecessary dict call - rewrite as a literal.
```

**Comment by amateja** [comment]
Same story for python 2.7.13. Only step 3 looks slightly different.
```
Package               Version
--------------------- -------
appdirs               1.4.3  
configparser          3.5.0  
enum34                1.1.6  
flake8                3.3.0  
flake8-comprehensions 1.4.0  
mccabe                0.6.1  
packaging             16.8   
pip                   9.0.1  
pycodestyle           2.3.1  
pyflakes              1.5.0  
pyparsing             2.2.0  
setuptools            35.0.2 
six                   1.10.0 
wheel                 0.29.0 
```

**Comment by amateja** [comment]
I would recommend to change your test accordingly:
```
def test_C408_pass_5():
    errors = run_flake8("""
        foo = {}
        bar = {'a': 1}
        baz = dict(foo=foo, **bar)
    """)
    assert errors == []
```
And it fails.

Difference between `baz = dict(foo, **bar)` and `baz = dict(foo=foo, **bar)` is substantial. First makes new dict from keys and values of `foo` and `bar`, second puts `foo` under `foo` key. Hope it helped.

**Comment by adamchainz** [comment]
Aha, of course, we missed the `dict(foo=1, **bar)` style. I have a minimal test case now:

```patch
diff --git a/tests/test_it.py b/tests/test_it.py
index de600c8..8c8c749 100644
--- a/tests/test_it.py
+++ b/tests/test_it.py
@@ -350,6 +350,14 @@ def test_C408_pass_4():
     assert errors == []
 
 
+def test_C408_pass_5():
+    errors = run_flake8('''\
+        foo = {}
+        dict(bar=1, **foo)
+    ''')
+    assert errors == []
+
+
 def test_C408_fail_1():
     errors = run_flake8('tuple()')
     assert errors == [

```

and am working on a fix.

**Comment by adamchainz** [comment]
Fix released in version 1.4.1 https://pypi.python.org/pypi/flake8-comprehensions

**Comment by graingert** [comment]
> This example could be rewritten to use literals, but it would be less succinct and (I assume, but haven't confirmed) less performant.

```python
def foo(data):
    defaults = {'a': 1, 'b': 2}
    return dict(defaults, **data)
```
```
  2           0 LOAD_CONST               1 (1)
              2 LOAD_CONST               2 (2)
              4 LOAD_CONST               3 (('a', 'b'))
              6 BUILD_CONST_KEY_MAP      2
              8 STORE_FAST               1 (defaults)

  3          10 LOAD_GLOBAL              0 (dict)
             12 LOAD_FAST                1 (defaults)
             14 BUILD_TUPLE              1
             16 LOAD_FAST                0 (data)
             18 CALL_FUNCTION_EX         1
             20 RETURN_VALUE
```

```python
def foo(data):
    return {"a": 1, "b": 2, **data}
```

```
  2           0 LOAD_CONST               1 (1)
              2 LOAD_CONST               2 (2)
              4 LOAD_CONST               3 (('a', 'b'))
              6 BUILD_CONST_KEY_MAP      2
              8 LOAD_FAST                0 (data)
             10 BUILD_MAP_UNPACK         2
             12 RETURN_VALUE
```

I'm assuming that BUILD_MAP_UNPACK is faster than LOAD_GLOBAL,LOAD_FAST,CALL_FUNCTION_EX

**Comment by adamchainz** [comment]
Looks like it is, for small dicts at least:

```
In [1]: def foo(data):
   ...:     defaults = {'a': 1, 'b': 2}
   ...:     return dict(defaults, **data)
   ...:

In [2]: %timeit foo({'c': 3})
397 ns ± 8.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: def foo(data):
   ...:     return {"a": 1, "b": 2, **data}
   ...:

In [4]: %timeit foo({'c': 3})
266 ns ± 4.57 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
